README.md
# Snapshooter

Simple crawler for Single Page Applications.

> Version 0.2.13

[![Dependency Status](https://gemnasium.com/serpentem/snapshooter.png)](https://gemnasium.com/serpentem/snapshooter)

# About

Snapshooter will load a URL, wait the javascript to render and save it as plain
HTML.

This will be done recursively until all found links in the start URL is saved.

# Issues

Do not hesitate to open a feature request or a bug report.
> https://github.com/serpentem/snapshooter/issues

# Docs
  - [Requirements](#requirements)
  - [Installing](#installing)
  - [Help](#help)
    - [Integration](#integration)
  - [Contributing](#contributing)

<a name="requirements" />
## Requirements

You'll need PhantomJS installed (v 1.9 or greater) in order to use this package.
 * http://phantomjs.org


<a name="installing" />
## Installing

````bash
npm install -g snapshooter
````

<a name="help" />
## Help

````
Usage:
  snapshooter [options] [params]

Options:
  -i, --input            Input url to index                                                 
  -o, --output           Output folder to save indexed files                                
  -e, --exclude          Regex pattern for excluding files (pass between quotes)            
  -p, --pretty           Output indexed files in a pretty fashion way                       
  -s, --server           Start a server for previewing indexed content                      
  -P, --port             Preview server port                                                  [default: 8080]
  -f, --forward          Avoid indexing links up to the initial url folder                  
  -t, --timeout          Time limit (in seconds) to wait for a page to render                 [default: 15]
  -m, --max-connections  Max connections limit, use with care                                 [default: 10]
  -l, --log              Show 'console.log' messages (try disabling it if phantom crashes)  
  -O, --once             Avoid recursivity, index only the given url and nothing else       
  -S, --stdout           Prints indexed content to stdout (auto-set -O=true -l=false)       
  -V, --verbose          Shows info logs about files skipped                                
  -D, --delete           Automatically delete destination folder before writing new files   
  -X, --overwrite        Automatically overwrite destination folder with new files          
  -H, --hidden           Doesn't inject the `window.snapshooter=true` on pages being indexed
  -v, --version          Shows snapshooter version                                          
  -h, --help             Shows this help screen                                             


Examples:
  snapshooter -i <site.com> -o <local-folder>
  snapshooter -i <site.com> -o <local-folder> -p
  snapshooter -i <site.com> -o <local-folder> -ps [-P 3000] [-e '/\.exe$/m'] [-t 20000]
````

<a name="integration" />
### Integration

A very tiny bit of integration is needed in order for it to effectively wait
until all javascript opterations is done, such as data loadings, template
rendering etc.

Considering you have a **Single Page Application** I bet you have also some
`render` method, and possibly another `in` and `out` too for handling
transitions.

Well, the only matter here is to inform **Snapshooter** that the page has finish
rendering. It's achieved by setting the property `window.crawler.is_rendered`.

````javascript
window.crawler.is_rendered = true
````

Snapshooter will keep waiting for the page until this variable gets `true` and
then the rendered DOM will be saved as a plain html file.


<a name="contributing"/>
## Contributing

Download the repo and have fun, pull requests are more than welcome.

### Setting up

````bash
  git clone git://github.com/serpentem/snapshooter.git
  cd snapshooter
  npm link
````

### Compiling

To build, just run:

````bash
make build
````

During develop you may prefer:

````bash
make watch
````

### Attention

Do not mess with version number.

# Powered by
 - [CoffeeScript](https://github.com/jashkenas/coffee-script)
 - [PhantomJS-node](https://github.com/sgentle/phantomjs-node)
 - [PhantomJS](http://phantomjs.org)

HISTORY.md
# Changelog

## 0.2.12 - 06/30/2013
 * Upgrading outdated dependencies

## 0.2.11 - 06/30/2013
 * Properly exiting process when no `-s` option is given

## 0.2.10 - 18/06/2013
 * Fixing validation rule when overwriting existent directory, making [A]bort
 option default in order to avoid disasters.

## 0.2.9 - 12/06/2013
 * Added huge performance improvement to ease the pain out of indexing large
 apps.

## 0.2.8 - 12/06/2013
 * Adding `--hidden` option to avoid injection of `window.snapshooter` variable
 on pages being crawled.

## 0.2.7 - 10/06/2013
 * Adding handy options `--delete` and `--overwrite`
 * Injecting `snapshooter=true` variable on window of pages before crawling to
 let apps make decisions based on this
 * Removing line breaks from saved indexed pages when `--pretty` is not informed

## 0.2.6 - 10/06/2013
 * Adding `--log` option to show page `console.log` calls in terminal while
 crawling pages (useful for debug).

## 0.2.5 - 04/06/2013
 * Excluding `mailto:` links
 * Handling absence of first slash on hrefs

## 0.2.4 - 03/06/2013
 * Automatically set -O to true when using -S to avoid errors
 * Handling validation at startup to consider options -o or -S as mandatory 

## 0.2.3 - 03/06/2013
 * Passing regexes between single/double quotes is now mandatory
 * Changing --timeout to use seconds instead of milliseconds

## 0.2.2 - 02/06/2013
 * Fixing and improving validations at startup one more time
 * Validating specific version of installed phantomjs to reduce problems

## 0.2.1 - 02/06/2013
 * Fixing validations at startup

## 0.2.0 - 02/06/2013
 * Snapshooter REVAMPED!
 * Switching phantomjs bridge library
 * Fixing and improving url parsers
 * Adding proper CLI helper with flexible options and flags
  * --input
  * --output
  * --exclude
  * --pretty
  * --server
  * --port
  * --forward
  * --timeout
  * --max-connections
  * --once
  * --stdout
  * --verbose
  * --version
  * --help
 * Adding simple static WebServer for previewing crawled content
 * Adding makefile targets for publishing and watching project
 * Publishing pure javasript again in the name of God
 * Adding `source-map-support` to keep error msgs as sane as possible

lib/core/live.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Crawler, Live, http;

  http = require('http');

  Crawler = require('./crawler');

  /*
    Http proxy for crawling live
  */


  module.exports = Live = (function() {
    function Live(the, cli) {
      var server,
        _this = this;
      this.the = the;
      this.cli = cli;
      server = http.createServer(function(req, res) {
        res.writeHead(200, {
          'Content-Type': 'text/html'
        });
        if (/\./.test(req.url)) {
          return res.end();
        }
        console.warn('trying to catch', _this.cli.argv.input + req.url);
        return new Crawler(_this.cli, _this.cli.argv.input + req.url, function(source) {
          return res.end(source);
        });
      });
      server.listen(this.cli.argv.port);
    }

    return Live;

  })();

}).call(this);

/*
//@ sourceMappingURL=live.map
*/


lib/core/shoot.js
// Generated by CoffeeScript 1.6.2
(function() {
  var Crawler, Shoot, connect, fs, fsu, path;

  fs = require('fs');

  path = require('path');

  fsu = require('fs-util');

  connect = require('connect');

  Crawler = require('./crawler');

  /*
    Instantiate a crawler for the first url,
    Crawler returns a source and "<a href=''>" links url
  
    The links url are filtered ( i.e. external links are not crawled ),
    and then written to disk
  */


  module.exports = Shoot = (function() {
    Shoot.prototype.crawled = {};

    Shoot.prototype.skipped_urls = {};

    Shoot.prototype.domain = null;

    Shoot.prototype.root_folder = null;

    Shoot.prototype.pending_urls = null;

    Shoot.prototype.connections = 0;

    Shoot.prototype.max_connections = null;

    Shoot.prototype.start_time = null;

    Shoot.prototype.exclude = null;

    Shoot.prototype.crawled_files_num = 0;

    Shoot.prototype.failed_files_num = 0;

    Shoot.prototype.skipped_files_num = 0;

    function Shoot(the, cli) {
      var all, err, first_url, flags, msg, reg, _ref;

      this.the = the;
      this.cli = cli;
      this.max_connections = this.cli.argv.m;
      this.pending_urls = [];
      if (this.cli.argv.exclude != null) {
        reg = /[\/'"]+(.+)(?:\/)([mgi]*)['"]*$/m;
        _ref = reg.exec(this.cli.argv.exclude), all = _ref[0], reg = _ref[1], flags = _ref[2];
        try {
          this.exclude = new RegExp(reg, flags);
          msg = ' • WARN'.bold + ' Exclude filter in use: ' + this.exclude;
          console.log(msg.yellow);
        } catch (_error) {
          err = _error;
          msg = ' • ERROR '.bold + err;
          console.log(msg.red);
          console.log('Did you inform your pattern between quotes?'.cyan);
          process.exit();
        }
      }
      if (!~this.cli.argv.input.indexOf('http')) {
        first_url = this.cli.argv.input = 'http://' + this.cli.argv.input;
      }
      first_url = first_url.replace(/\/index\.\w+$/m, '');
      first_url = first_url.replace(/\/+$/m, '');
      this.domain = (first_url.match(/https?:\/\/[^\/]+/))[0];
      this.root_folder = first_url;
      this.start_time = (new Date).getTime();
      this.crawl(first_url);
    }

    Shoot.prototype.crawl = function(url) {
      var _this = this;

      if (this.crawled[url] !== void 0) {
        return;
      }
      this.crawled[url] = false;
      if (!this.cli.argv.stdout) {
        console.log('>'.bold.yellow, url.grey);
      }
      this.connections++;
      return new Crawler(this.cli, url, function(source) {
        if (!_this.cli.argv.stdout) {
          console.log('< '.bold.cyan, url.grey);
        }
        _this.connections--;
        _this.crawled[url] = true;
        if (source != null) {
          _this.crawled_files_num++;
        } else {
          _this.failed_files_num++;
        }
        if (_this.cli.argv.stdout) {
          console.log(source);
          return;
        }
        _this.save_page(url, source);
        if (_this.cli.argv.once) {
          return _this.finish();
        } else {
          return _this.after_crawl(source);
        }
      });
    };

    Shoot.prototype.after_crawl = function(source) {
      var absolute, links, match, passed, reg, relative, skipped;

      reg = /<a\s+href\s*=\s*["']+(?!http)([^"']+)/g;
      links = [];
      if (source != null) {
        while ((match = reg.exec(source)) != null) {
          relative = match[1];
          absolute = this.domain + ("/" + relative).replace(/^\/\//, '/');
          passed = this.url_is_passing(relative, absolute);
          if (passed) {
            this.pending_urls.push(absolute);
          } else {
            skipped = passed !== -1;
            skipped && (skipped = this.skipped_urls[absolute] === void 0);
            if (skipped) {
              this.skipped_urls[absolute] = true;
              this.skipped_files_num++;
            }
            if (skipped && this.cli.argv.verbose) {
              console.log('• INFO '.bold.cyan + ("skipping " + absolute.yellow).cyan);
            }
          }
        }
      }
      while (this.connections < this.max_connections && this.pending_urls.length) {
        this.crawl(this.pending_urls.shift());
      }
      if (this.connections === 0) {
        return this.finish();
      }
    };

    Shoot.prototype.save_page = function(url, source) {
      var output_file, output_folder, relative_url;

      relative_url = (url.replace(this.domain, '')) || '/';
      output_folder = path.join(this.cli.argv.output, relative_url);
      output_file = path.join(output_folder, 'index.html');
      if (!fs.existsSync(output_folder)) {
        fsu.mkdir_p(output_folder);
      }
      fs.writeFileSync(output_file, source);
      return console.log('✓ '.green, relative_url);
    };

    Shoot.prototype.finish = function() {
      var address, ms, root;

      Crawler.kill();
      if (!this.cli.argv.stdout) {
        ms = ((new Date).getTime() - this.start_time) + ' ms';
        console.log(("\n★  Application crawled successfully in " + ms.magenta).green);
        console.log('\t Indexed: ' + this.crawled_files_num);
        console.log('\t Skipped: ' + this.skipped_files_num);
        console.log('\t Failed: ' + this.failed_files_num);
      }
      if (!this.cli.argv.server) {
        return process.exit();
      }
      this.conn = connect().use(connect["static"](this.cli.argv.output)).listen(this.cli.argv.port);
      address = 'http://localhost:' + this.cli.argv.port;
      root = this.cli.argv.output;
      return console.log(("\nPreview server started for " + root + " at: \n\t").grey, address);
    };

    Shoot.prototype.url_is_passing = function(relative, absolute) {
      var flag, flags, not_anchor, not_backwards, not_crawled, not_exclude, not_image, not_mail_to, not_pdf, not_slash, not_zip, passed, _i, _len;

      not_slash = absolute !== '/';
      not_crawled = this.crawled[absolute] === void 0;
      not_anchor = relative !== '#';
      not_image = !(/\.(jpg|jpeg|gif|png)$/m.test(absolute));
      not_zip = !(/\.(zip|tar(\.gz)?)$/m.test(absolute));
      not_pdf = !(/\.(pdf)$/m.test(absolute));
      not_mail_to = !(/^mailto:/m.test(relative));
      if (this.cli.argv.forward) {
        not_backwards = (absolute.indexOf(this.root_folder)) === 0;
      } else {
        not_backwards = true;
      }
      if (this.exclude != null) {
        not_exclude = !(this.exclude.test(absolute));
      } else {
        not_exclude = true;
      }
      flags = [not_slash, not_crawled, not_anchor, not_image, not_zip, not_pdf, not_exclude, not_backwards, not_mail_to];
      passed = true;
      for (_i = 0, _len = flags.length; _i < _len; _i++) {
        flag = flags[_i];
        passed && (passed = flag);
      }
      if (passed) {
        return true;
      } else {
        if (not_crawled === false) {
          return -1;
        } else {
          return false;
        }
      }
    };

    return Shoot;

  })();

}).call(this);

/*
//@ sourceMappingURL=shoot.map
*/


lib/snapshooter.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Cli, Live, Shoot, Snapshooter, colors, exec, fs, fsu, path;

  require('source-map-support').install();

  fs = require('fs');

  path = require('path');

  colors = require('colors');

  exec = (require('child_process')).exec;

  fsu = require('fs-util');

  Shoot = require('./core/shoot');

  Live = require('./core/live');

  Cli = require('./cli');

  module.exports = Snapshooter = (function() {
    function Snapshooter() {
      var _this = this;
      this.version = (require('./../package.json')).version;
      this.cli = new Cli(this.version);
      if (this.cli.argv.version) {
        return console.log(this.version);
      }
      if (this.cli.argv.input) {
        return this.check_phantom(function() {
          return _this.init();
        });
      }
      console.log(this.cli.opts.help() + this.cli.examples);
    }

    Snapshooter.prototype.init = function() {
      var live, msg,
        _this = this;
      console.warn('live is :', this.cli.argv.live);
      console.warn('port is :', this.cli.argv.port);
      if (this.cli.argv.live) {
        live = new Live(this["this"], this.cli);
        return;
      }
      if (!(this.cli.argv.output || this.cli.argv.stdout)) {
        console.log('• ERROR '.bold.red + 'Output dir not informed!');
        console.log(' ➜  Alternatively, you can pass -S to write to stdout.'.cyan);
        return;
      }
      if (fs.existsSync(this.cli.argv.output)) {
        if (this.cli.argv["delete"] != null) {
          fsu.rm_rf(this.cli.argv.output);
          return this.shoot();
        } else if (this.cli.argv.overwrite != null) {
          return this.shoot();
        } else {
          msg = "Output folder exists:\n  [O]verwrite existent files\n  [D]elete folder before writing new files\n  [A]bort\n\nEnter option";
          return this.prompt(msg, /(o|d|a)/i, 'A', function(answer) {
            answer = answer.toLowerCase();
            if (answer.toLowerCase() === 'd') {
              fsu.rm_rf(_this.cli.argv.output);
              return _this.shoot();
            } else if (answer.toLowerCase() === 'o') {
              return _this.shoot();
            } else {
              return process.exit();
            }
          });
        }
      } else {
        return this.shoot();
      }
    };

    Snapshooter.prototype.shoot = function() {
      return new Shoot(this, this.cli);
    };

    Snapshooter.prototype.check_phantom = function(fn) {
      return this.has_phantom(function(status) {
        var msg;
        if (status === false) {
          msg = ('• ERROR'.bold + ' First you must to install ').red;
          msg += 'phantomjs v1.9.0'.yellow + ' or greater '.red;
          msg += '\n\t➜  http://phantomjs.org'.cyan;
          console.error(msg);
          return process.exit();
        }
        return fn();
      });
    };

    Snapshooter.prototype.has_phantom = function(fn) {
      var _this = this;
      return exec("phantomjs -v", function(error, stdout, stderr) {
        return fn(/1\.9\.[0-9]+/.test(stdout));
      });
    };

    Snapshooter.prototype.prompt = function(question, format, _default, fn) {
      var stdin, stdout,
        _this = this;
      stdin = process.stdin;
      stdout = process.stdout;
      if (_default != null) {
        question += " [" + _default + "]";
      }
      stdout.write("" + question + ": ");
      return stdin.once('data', function(data) {
        var msg, rule;
        data = data.toString().trim() || _default;
        if (format.test(data)) {
          return fn(data.trim());
        } else {
          msg = "" + 'Invalid entry, it should match:'.red;
          rule = "" + (format.toString().cyan);
          stdout.write("\t" + msg + " " + rule + "\n");
          return _this.prompt(question, format, fn);
        }
      }).resume();
    };

    return Snapshooter;

  })();

}).call(this);

/*
//@ sourceMappingURL=snapshooter.map
*/


lib/core/crawler.js
// Generated by CoffeeScript 1.6.2
(function() {
  var Crawler, pd, phantom,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  phantom = require('phantom');

  pd = (require('pretty-data')).pd;

  /*
    Create and instantiate phantomjs thought phantom
  */


  module.exports = Crawler = (function() {
    var ph, port;

    ph = null;

    port = 12345;

    Crawler.prototype.page = null;

    Crawler.prototype.start_time = null;

    port = 12345;

    function Crawler(cli, url, done) {
      var _this = this;

      this.cli = cli;
      this.url = url;
      this.done = done;
      this.keep_on_checking = __bind(this.keep_on_checking, this);
      this.start_time = (new Date).getTime();
      this.create_phantom(function() {
        return _this.create_page(function() {
          return _this.open_url();
        });
      });
    }

    Crawler.kill = function() {
      return ph.exit();
    };

    Crawler.prototype.create_phantom = function(done) {
      var _this = this;

      if (ph != null) {
        return done();
      }
      return phantom.create('--load-images=false', function(_ph, err) {
        var msg;

        if (err != null) {
          msg = '• ERROR '.bold.red + 'phantom.create couldn\'t finish for '.red;
          msg += _this.url.yellow;
          return _this.error(msg, err);
        }
        ph = _ph;
        return done();
      }, 'phantomjs', port++);
    };

    Crawler.prototype.create_page = function(done) {
      var _this = this;

      return ph.createPage(function(page, err) {
        var msg;

        _this.page = page;
        if (err != null) {
          msg = '• ERROR '.bold.red + 'ph.createPage couldn\'t finish for '.red;
          msg += _this.url.yellow;
          return _this.error(msg, err);
        }
        if (_this.cli.argv.log && !_this.cli.argv.stdout) {
          _this.page.set('onConsoleMessage', function(msg) {
            return console.log(msg);
          });
        }
        if (!_this.cli.argv.hidden) {
          _this.page.set('onInitialized', function() {
            return _this.page.evaluate(function() {
              return window.snapshooter = true;
            });
          });
        }
        return done();
      });
    };

    Crawler.prototype.open_url = function() {
      var _this = this;

      return this.page.open(this.url, function(status, err) {
        var msg;

        if (err != null) {
          msg = '• ERROR '.bold.red + 'page.open couldn\'t finish for '.red;
          msg += _this.url.yellow;
          return _this.error(msg, err);
        }
        if (status !== 'success') {
          msg = '• ERROR '.bold.red + ' ' + _this.url.yellow;
          msg += " ended with status = ".red + status.bold;
          return _this.error(msg);
        }
        return _this.keep_on_checking();
      });
    };

    Crawler.prototype.keep_on_checking = function() {
      var _this = this;

      return this.page.evaluate(function() {
        var data;

        return data = {
          rendered: window.crawler && window.crawler.is_rendered,
          source: document.all[0].outerHTML
        };
      }, function(data) {
        var msg;

        if (!(data != null ? data.rendered : void 0)) {
          if (((new Date).getTime()) - _this.start_time > (_this.cli.argv.timeout * 1000)) {
            msg = '• ERROR '.bold.red + _this.url.yellow;
            msg += ' took too long to render, skipping'.red;
            return _this.error(msg);
          } else {
            return setTimeout(_this.keep_on_checking, 100);
          }
        }
        if (_this.cli.argv.pretty) {
          return _this.done(pd.xml(data.source));
        } else {
          return _this.done(data.source.replace(/\n/g, ''));
        }
      });
    };

    Crawler.prototype.error = function(msg, error) {
      console.error(msg);
      if (error != null) {
        console.error(error);
      }
      ph.exit();
      return this.done(null);
    };

    return Crawler;

  })();

}).call(this);

/*
//@ sourceMappingURL=crawler.map
*/


lib/cli.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Cli, optimist;

  optimist = require('optimist');

  module.exports = Cli = (function() {
    function Cli(version) {
      this.version = version;
      this.usage = ("" + 'Snapshooter' + " ") + ("v" + this.version + "\n").grey;
      this.usage += "Simple crawler for Single Page Applications\n\n".grey;
      this.usage += "" + 'Usage:' + "\n";
      this.usage += "  snapshooter [" + 'options'.green + "] [" + 'params'.green + "]";
      this.examples = "\n\nExamples:\n  snapshooter -i <site.com> -o <local-folder>\n  snapshooter -i <site.com> -o <local-folder> -p\n  snapshooter -i <site.com> -o <local-folder> -ps [-P 3000] [-e '/\\.exe$/m'] [-t 20000]\n";
      this.argv = (this.opts = optimist.usage(this.usage).alias('i', 'input').describe('i', 'Input url to index').alias('o', 'output').describe('o', 'Output folder to save indexed files').alias('e', 'exclude').describe('e', 'Regex pattern for excluding files (pass between quotes)').alias('p', 'pretty').describe('p', 'Output indexed files in a pretty fashion way').alias('s', 'server').describe('s', 'Start a server for previewing indexed content').alias('P', 'port').describe('P', 'Preview server port')["default"]('P', 8080).alias('f', 'forward').describe('f', 'Avoid indexing links up to the initial url folder').alias('t', 'timeout').describe('t', 'Time limit (in seconds) to wait for a page to render')["default"]('t', 15).alias('m', 'max-connections').describe('m', 'Max connections limit, use with care')["default"]('m', 10).alias('l', 'log').describe('l', 'Show \'console.log\' messages (try disabling it if phantom crashes)').alias('L', 'live').describe('L', 'creates a "live" tunnel, crawling will happen on demand ').alias('O', 'once').describe('O', 'Avoid recursivity, index only the given url and nothing else').alias('S', 'stdout').describe('S', 'Prints indexed content to stdout (auto-set -O=true -l=false)').alias('V', 'verbose').describe('V', 'Shows info logs about files skipped').alias('D', 'delete').describe('D', 'Automatically delete destination folder before writing new files').alias('X', 'overwrite').describe('X', 'Automatically overwrite destination folder with new files').alias('H', 'hidden').describe('H', 'Doesn\'t inject the `window.snapshooter=true` on pages being indexed').alias('v', 'version').describe('v', 'Shows snapshooter version').alias('h', 'help').describe('h', 'Shows this help screen')).argv;
    }

    return Cli;

  })();

}).call(this);

/*
//@ sourceMappingURL=cli.map
*/
